swarm.js

```diff
@@ -1,38 +1,146 @@
 'use strict'
 const peerInfo = require('./lib/peer-info')
+const peerQueue = require('./lib/queue')
 const { EventEmitter } = require('events')
 const guts = require('@hyperswarm/guts')
 
+const MAX_PEERS_DEFAULT = 16
+const ERR_MISSING_KEY = 'key is required and must be a buffer'
+
 module.exports = opts => new Swarm(opts)
 
 class Swarm extends EventEmitter {
   constructor (opts = {}) {
     super()
-    const { bootstrap, ephemeral } = opts
-    this.network = guts({
+    const {
+      maxPeers = MAX_PEERS_DEFAULT,
+      bootstrap,
+      ephemeral
+    } = opts
+    const queue = peerQueue()
+    const network = guts({
       bootstrap,
       ephemeral,
-      bind: () => {
-        this.emit('listening')
-      },
+      bind: () => this.emit('listening'),
       socket: (socket, isTCP) => {
         const info = peerInfo(null)
         info.connected(socket, isTCP)
         this.emit('connection', socket, info)
       },
-      close: () => {
-        this.emit('close')
-      }
+      close: () => this.emit('close')
     })
+    queue.on('readable', this._drain(queue))
+    this.peers = 0
+    this.maxPeers = maxPeers
+    this.emphemeral = ephemeral !== false
+    this.network = network
+    this.queue = queue
+  }
+  _drain (queue) {
+    const onConnect = (info) => (err, socket, isTCP) => {
+      if (err) {
+        this.peers -= 1
+        if (!queue.requeue(info)) queue.requeue(info)
+        drain()
+        return
+      }
+      if (info.reconnecting) info.reconnect(false)
+      info.connected(socket, isTCP)
+      this.emit('connection', socket, info)
+      socket.on('close', () => {
+        this.peers -= 1
+        info.disconnected()
+        info.reconnect(true)
+        if (!queue.requeue(info)) queue.requeue(info)
+      })
+      drain()
+    }
+    const drain = () => {
+      const info = queue.shift()
+      if (info && this.peers < this.maxPeers) {
```

Comment:
we want to make move this peers < maxPeers check up before the queue shift

Reply:
of course 🤦‍♂